test/cmocka/src/math/dct/dct.c

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <errno.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <string.h>
+#include <cmocka.h>
+#include <math.h>
+#include <sof/math/dct.h>
+#include <sof/math/matrix.h>
+#include "ref_dct_matrix_16_test1.h"
+#include "ref_dct_matrix_16_test2.h"
+
+#define MATRIX_MULT_16_MAX_ERROR_ABS  2.5
+#define MATRIX_MULT_16_MAX_ERROR_RMS  1.1
+
+static void dct_matrix_16_test(const int16_t *ref, int num_in, int num_out,
+			       enum dct_type type, bool ortho)
+{
+	struct dct_plan_16 dct;
+	float delta;
+	float sum_squares = 0;
+	float error_rms;
+	float delta_max = 0;
+	int16_t x;
+	int rows;
+	int columns;
+	int ret;
+	int i, j, k;
+
+	/* Calculate DCT matrix */
+	dct.num_in = num_in;
+	dct.num_out = num_out;
+	dct.type = type;
+	dct.ortho = ortho;
+	ret = dct_initialize_16(&dct);
+	if (ret) {
+		fprintf(stderr, "Failed to initialize DCT.\n");
+		exit(EXIT_FAILURE);
+	}
+
+	/* Check */
+	rows = dct.matrix->rows;
+	columns = dct.matrix->columns;
+	k = 0;
+	for (i = 0; i < rows; i++) {
+		for (j = 0; j < columns; j++) {
+			x = mat_get_scalar_16b(dct.matrix, i, j);
+			delta = (float)x - (float)ref[k++];
```

Comment:
Choosing `double` rather than `float` may be preferable for `x`.

Reply:
Yep good idea, especially if we add 32 bit version.